File: rust/src/parser.rs
use core::panic;

use crate::model::{Expression, Node, Statement, Statements};
use crate::token::{Token, TokenType};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Parser {
        Parser { tokens, current: 0 }
    }

    fn advance(&mut self) -> Option<Token> {
        if self.current < self.tokens.len() {
            let token = self.tokens[self.current].clone();
            self.current += 1;
            return Some(token);
        }
        None
    }
    fn is_next(&self, expected_type: TokenType) -> bool {
        match self.peek() {
            Some(token) => token.token_type == expected_type,
            None => false,
        }
    }

    fn expect(&mut self, expected_type: TokenType) -> Token {
        match self.peek() {
            Some(token) => {
                if token.token_type == expected_type {
                    return self.advance().unwrap();
                }
                panic!("Unexpected token")
            }
            None => panic!("Unexpected token"),
        }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.current)
    }

    fn match_token(&mut self, token_type: TokenType) -> bool {
        match self.peek() {
            Some(token) => {
                if token.token_type == token_type {
                    self.current += 1;
                    return true;
                }
                false
            }
            None => false,
        }
    }

    fn previous_token(&self) -> Option<Token> {
        self.tokens.get(self.current - 1).cloned()
    }

    fn exponent(&mut self) -> Expression {
        let mut expr = self.factor();
        while self.match_token(TokenType::TokCaret) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.exponent()),
                    }
                }
                None => panic!("Couldn't get previous_token"),
            };
        }
        expr
    }

    fn modulo(&mut self) -> Expression {
        let mut expr = self.exponent();
        while self.match_token(TokenType::TokMod) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.factor()),
                    }
                }
                None => panic!("Couldn't get previous_token"),
            };
        }
        expr
    }

    fn term(&mut self) -> Expression {
        let mut expr = self.modulo();
        while self.match_token(TokenType::TokStar) || self.match_token(TokenType::TokSlash) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.modulo()),
                    }
                }
                None => panic!("Couldn't get previous_token"),
            };
        }
        expr
    }

    fn expr(&mut self) -> Expression {
        let mut expr = self.term();
        while self.match_token(TokenType::TokPlus) || self.match_token(TokenType::TokMinus) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.term()),
                    }
                }
                None => panic!("Failed to get a previous_token"),
            };
        }
        expr
    }

    fn primary(&mut self) -> Expression {
        if self.match_token(TokenType::TokInteger) {
            match self.previous_token() {
                Some(token) => {
                    return Expression::Integer {
                        value: token.lexeme.parse::<i64>().unwrap(),
                    }
                }
                None => panic!("Failed getting Integer"),
            };
        };
        if self.match_token(TokenType::TokFloat) {
            match self.previous_token() {
                Some(token) => {
                    return Expression::Float {
                        value: token.lexeme.parse::<f64>().unwrap(),
                    }
                }
                None => panic!("Failed getting Float"),
            };
        };
        if self.match_token(TokenType::TokLparen) {
            let expr = self.logical_or();
            if self.match_token(TokenType::TokRparen) {
                return Expression::Grouping {
                    value: Box::new(expr),
                };
            }
            panic!("Unable to find closing parentheses");
        }
        if self.match_token(TokenType::TokTrue) {
            match self.previous_token() {
                Some(_) => return Expression::Bool { value: true },
                _ => panic!("Unable to parse primary token"),
            }
        }
        if self.match_token(TokenType::TokFalse) {
            match self.previous_token() {
                Some(_) => return Expression::Bool { value: false },
                _ => panic!("Unable to parse primary token"),
            }
        }
        if self.match_token(TokenType::TokString) {
            match self.previous_token() {
                Some(token) => {
                    let len = token.lexeme.len();
                    return Expression::Str {
                        value: token.lexeme[1..len - 1].to_string(),
                    };
                }
                _ => panic!("Unable to parse primary token"),
            }
        }
        Expression::Identifier {
            name: self.expect(TokenType::TokIdentifier).lexeme,
        }
    }

    fn unary(&mut self) -> Expression {
        if self.match_token(TokenType::TokNot)
            || self.match_token(TokenType::TokMinus)
            || self.match_token(TokenType::TokPlus)
        {
            match self.previous_token() {
                Some(token) => {
                    return Expression::UnaryOp {
                        op: token,
                        exp: Box::new(self.unary()),
                    }
                }
                None => panic!("No previous_token"),
            };
        };
        self.primary()
    }

    fn factor(&mut self) -> Expression {
        self.unary()
    }

    fn comparison(&mut self) -> Expression {
        let mut expr = self.expr();
        while self.match_token(TokenType::TokGe)
            || self.match_token(TokenType::TokLe)
            || self.match_token(TokenType::TokGt)
            || self.match_token(TokenType::TokLt)
        {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.expr()),
                    }
                }
                _ => panic!("No previous_token"),
            }
        }
        expr
    }

    fn equality(&mut self) -> Expression {
        let mut expr = self.comparison();
        while self.match_token(TokenType::TokEq) || self.match_token(TokenType::TokNe) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::BinOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.comparison()),
                    }
                }
                None => panic!("No previous_token"),
            };
        }
        expr
    }

    fn logical_and(&mut self) -> Expression {
        let mut expr = self.equality();
        while self.match_token(TokenType::TokAnd) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::LogicalOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.equality()),
                    }
                }
                None => panic!("No previous_token"),
            };
        }
        expr
    }

    fn logical_or(&mut self) -> Expression {
        let mut expr = self.logical_and();
        while self.match_token(TokenType::TokOr) {
            match self.previous_token() {
                Some(op) => {
                    expr = Expression::LogicalOp {
                        op,
                        left: Box::new(expr),
                        right: Box::new(self.logical_and()),
                    }
                }
                None => panic!("No previous_token"),
            };
        }
        expr
    }

    fn if_stmt(&mut self) -> Statement {
        self.expect(TokenType::TokIf);
        let test = self.logical_or();
        self.expect(TokenType::TokThen);
        let then_stmts = self.stmts();
        let mut else_stmts: Statements = Vec::new();
        if self.is_next(TokenType::TokElse) {
            self.advance();
            else_stmts = self.stmts();
        }
        self.expect(TokenType::TokEnd);
        Statement::IfStatement {
            test,
            then_stmts,
            else_stmts,
        }
    }

    fn while_stmt(&mut self) -> Statement {
        self.expect(TokenType::TokWhile);
        let test = self.logical_or();
        self.expect(TokenType::TokThen);
        let stmts = self.stmts();
        self.expect(TokenType::TokEnd);
        return Statement::While { test, stmts };
    }

    fn println_stmt(&mut self) -> Statement {
        if self.match_token(TokenType::TokPrintln) {
            return Statement::PrintlnStatement {
                value: self.logical_or(),
            };
        }
        panic!("Wrong token type")
    }

    fn print_stmt(&mut self) -> Statement {
        if self.match_token(TokenType::TokPrint) {
            return Statement::PrintStatement {
                value: self.logical_or(),
            };
        }
        panic!("Wrong token type")
    }

    fn stmt(&mut self) -> Statement {
        let token = self.peek();
        match token {
            Some(op) => match op.token_type {
                TokenType::TokPrint => self.print_stmt(),
                TokenType::TokPrintln => self.println_stmt(),
                TokenType::TokWhile => self.while_stmt(),
                TokenType::TokIf => self.if_stmt(),
                _ => {
                    let left = self.expr();
                    if self.match_token(TokenType::TokAssign) {
                        return Statement::Assignment {
                            left,
                            right: self.expr(),
                        };
                    }
                    panic!("Not handling function calls yet");
                }
            },
            None => panic!("Failed to get token"),
        }
    }

    fn stmts(&mut self) -> Statements {
        let mut stmts_vec: Statements = Vec::new();
        while self.current < self.tokens.len()
            && !self.is_next(TokenType::TokEnd)
            && !self.is_next(TokenType::TokElse)
        {
            stmts_vec.push(self.stmt());
        }
        stmts_vec
    }

    pub fn parse(&mut self) -> Node {
        Node::Stmts(self.stmts())
    }
}

File: c/parser.c
#include "parser.h"
#include "memory.h"
#include "model.h"
#include "tokens.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/types.h>

Expression *push_expression(Parser *self, Expression expr) {
  Expression *new_expression = arena_alloc(self->arena, sizeof(Expression));
  *new_expression = expr;
  return new_expression;
}

Token *advance_parser(Parser *self) {
  if (self->current < self->tokens_list_len) {
    return ((Token *)self->arena->memory) + self->current++;
  }
  assert("Shouldn't request more then len token");
  return NULL;
}

int is_next(Parser *self, TokenType expected_type) {
  Token token = peek_token(self);
  if (token.token_type == expected_type)
    return 1;
  return 0;
}

Token *expect(Parser *self, TokenType expected_type) {
  Token token = peek_token(self);
  if (token.token_type == expected_type) {
    return advance_parser(self);
  };
  return NULL;
}

Token peek_token(Parser *self) {
  if (self->current < self->tokens_list_len) {
    return ((Token *)self->arena->memory)[self->current];
  }
  assert("Shouldn't request more then len token");
  return (Token){};
}

int match_token(Parser *self, TokenType expected_type) {
  Token token = peek_token(self);
  self->current += token.token_type == expected_type;
  return token.token_type == expected_type;
}

Token previous_token(Parser *self) {
  if (self->current > 0)
    return ((Token *)self->arena->memory)[self->current - 1];
  assert("Shouldn't request -1 token");
  return (Token){};
}

Expression *exponent(Parser *self) {
  Expression *express = factor(self);
  while (match_token(self, TokCaret)) {
    express = push_expression(
        self, (Expression){BINARY_OP, .BinaryOp = {previous_token(self),
                                                   express, exponent(self)}});
  }
  return express;
}

Expression *modulo(Parser *self) {
  Expression *express = exponent(self);
  while (match_token(self, TokMod)) {
    express = push_expression(
        self, (Expression){BINARY_OP, .BinaryOp = {previous_token(self),
                                                   express, factor(self)}});
  }
  return express;
}

Expression *term(Parser *self) {
  Expression *express = modulo(self);
  while (match_token(self, TokStar) || match_token(self, TokSlash)) {
    express = push_expression(
        self, (Expression){BINARY_OP, .BinaryOp = {previous_token(self),
                                                   express, modulo(self)}});
  }
  return express;
}

Expression *expr(Parser *self) {
  Expression *express = term(self);
  while (match_token(self, TokPlus) || match_token(self, TokMinus)) {
    express = push_expression(
        self, (Expression){BINARY_OP, .BinaryOp = {previous_token(self),
                                                   express, term(self)}});
  }
  return express;
}

Expression *primary(Parser *self) {
  if (match_token(self, TokInteger)) {
    Token token = previous_token(self);
    return push_expression(
        self,
        (Expression){INTEGER, .Integer = {strtol(token.lexeme, NULL, 10)}});
  }
  if (match_token(self, TokFloat)) {
    Token token = previous_token(self);
    return push_expression(
        self, (Expression){FLOAT, .Float = {strtof(token.lexeme, NULL)}});
  }
  if (match_token(self, TokTrue)) {
    return push_expression(self, (Expression){BOOL, .Bool = {1}});
  }
  if (match_token(self, TokFalse)) {
    return push_expression(self, (Expression){BOOL, .Bool = {0}});
  }
  if (match_token(self, TokString)) {
    Token token = previous_token(self);
    return push_expression(self, (Expression){STRING, .String = {
                                                          token.lexeme + 1,
                                                          token.lexeme_len - 2,
                                                      }});
  }
  if (match_token(self, TokLparen)) {
    Expression *express = logical_or(self);
    {
      if (match_token(self, TokRparen)) {
        return push_expression(self,
                               (Expression){GROUPING, .Grouping = {express}});
      }
    }
  }
  Token *token = expect(self, TokIdentifier);
  return push_expression(
      self, (Expression){.type = IDENTIFIER,
                         .Identifier = {.name = token->lexeme,
                                        .len = token->lexeme_len}});
}

Expression *unary(Parser *self) {
  if (match_token(self, TokNot) || match_token(self, TokMinus) ||
      match_token(self, TokPlus)) {
    return push_expression(
        self,
        (Expression){UNARY_OP, .UnaryOp = {previous_token(self), unary(self)}});
  }
  return primary(self);
}

Expression *factor(Parser *self) { return unary(self); }

Expression *compare(Parser *self) {
  Expression *express = expr(self);
  while (match_token(self, TokGt) || match_token(self, TokLt) ||
         match_token(self, TokGe) || match_token(self, TokLe)) {
    express = push_expression(
        self, (Expression){LOGICAL_OP, .LogicalOp = {previous_token(self),
                                                     express, expr(self)}});
  }
  return express;
}

Expression *equality(Parser *self) {
  Expression *express = compare(self);
  while (match_token(self, TokEq) || match_token(self, TokNe)) {
    express = push_expression(
        self, (Expression){LOGICAL_OP, .LogicalOp = {previous_token(self),
                                                     express, compare(self)}});
  }
  return express;
}

Expression *logical_and(Parser *self) {
  Expression *express = equality(self);
  while (match_token(self, TokAnd)) {
    express = push_expression(
        self, (Expression){LOGICAL_OP, .LogicalOp = {previous_token(self),
                                                     express, equality(self)}});
  }
  return express;
}

Expression *logical_or(Parser *self) {
  Expression *express = logical_and(self);
  while (match_token(self, TokOr)) {
    express = push_expression(
        self,
        (Expression){LOGICAL_OP, .LogicalOp = {previous_token(self), express,
                                               logical_and(self)}});
  }
  return express;
}

Statement if_stmt(Parser *self) {
  expect(self, TokIf);
  Expression *test = logical_or(self);
  expect(self, TokThen);
  Statements *then_stmts = stmts(self);
  Statements *else_stmts = arena_alloc(self->arena, sizeof(Statements));
  if (is_next(self, TokElse)) {
    advance_parser(self);
    else_stmts = stmts(self);
  }
  expect(self, TokEnd);
  return (Statement){.type = IF, .IfStatement = {test, then_stmts, else_stmts}};
}

Statement while_stmt(Parser *self) {
  expect(self, TokWhile);
  Expression *test = logical_or(self);
  expect(self, TokThen);
  Statements *while_stmts = stmts(self);
  expect(self, TokEnd);
  return (Statement){.type = WHILE,
                     .While = {.test = test, .stmts = while_stmts}};
}

Statement println_stmt(Parser *self) {
  if (match_token(self, TokPrintln)) {
    return (Statement){.type = PRINTLN,
                       .PrintlnStatement.value = logical_or(self)};
  }
  exit(10);
}

Statement print_stmt(Parser *self) {
  if (match_token(self, TokPrint)) {
    return (Statement){.type = PRINT, .PrintStatement.value = logical_or(self)};
  }
  exit(11);
}

Statement for_stmt(Parser *self) {
  expect(self, TokFor);
  Expression *identifier = primary(self);
  expect(self, TokAssign);
  Expression *start = logical_or(self);
  expect(self, TokComma);
  Expression *stop = logical_or(self);
  Expression *step =
      push_expression(self, (Expression){.type = INTEGER, .Integer = {1}});
  if (match_token(self, TokComma)) {
    step = expr(self);
  }
  expect(self, TokDo);
  Statements *for_stmts = stmts(self);
  expect(self, TokEnd);
  return (Statement){.type = FOR,
                     .For = {.identifier = identifier,
                             .start = start,
                             .stop = stop,
                             .step = step,
                             .stmts = for_stmts}};
}

Statement stmt(Parser *self) {
  Token tok = peek_token(self);
  switch (tok.token_type) {
  case TokPrint:
    return print_stmt(self);
  case TokPrintln:
    return println_stmt(self);
  case TokIf:
    return if_stmt(self);
  case TokWhile:
    return while_stmt(self);
  case TokFor:
    return for_stmt(self);
  default:;
    Expression *left = primary(self);
    if (match_token(self, TokAssign)) {
      Expression *right = expr(self);
      return (Statement){.type = ASSIGNMENT,
                         .Assignment = {.left = left, .right = right}};
    }
    assert("Unknown type of statemnt");
  }
  return (Statement){};
}

Statements *stmts(Parser *self) {
  Statement *curr = arena_alloc(self->arena, sizeof(Statement));
  Statements *stmts_arr = arena_alloc(self->arena, sizeof(Statements));
  stmts_arr->head = curr;
  while (true) {
    *curr = stmt(self);
    if (!((self->current < self->tokens_list_len - 1) &&
          (!is_next(self, TokElse)) && (!is_next(self, TokEnd))))
      break;
    curr->next = arena_alloc(self->arena, sizeof(Statement));
    curr = curr->next;
  };
  return stmts_arr;
}

Node parse(Parser *self) {
  return (Node){.type = STMTS, .stmts = stmts(self)};
};
